client/my-sites/site-settings/hosting/controller.js

```diff
@@ -0,0 +1,56 @@
+/**
+ * External dependencies
+ */
+import React from 'react';
+import page from 'page';
+
+/**
+ * Internal Dependencies
+ */
+import Hosting from './main';
+import canSiteViewAtomicHosting from 'state/selectors/can-site-view-atomic-hosting';
+import { getCurrentPlan } from 'state/sites/plans/selectors';
+import { getSelectedSiteId } from 'state/ui/selectors';
+import { fetchSitePlans } from 'state/sites/plans/actions';
+
+function waitForData( context ) {
+	return new Promise( resolve => {
+		const unsubscribe = context.store.subscribe( () => {
+			const state = context.store.getState();
+
+			const siteId = getSelectedSiteId( state );
+			if ( ! siteId ) {
+				return;
+			}
+
+			const currentPlan = getCurrentPlan( state, siteId );
+			if ( ! currentPlan ) {
+				return;
+			}
+			unsubscribe();
+			resolve();
+		} );
+		// Trigger a `store.subscribe()` callback
+		context.store.dispatch( fetchSitePlans( getSelectedSiteId( context.store.getState() ) ) );
+	} );
+}
+
+export async function handleHostingPanelRedirect( context, next ) {
+	const { store } = context;
+
+	await waitForData( context );
```

Comment:
This is to ensure the plan is set on the state before `canSiteViewAtomicHosting` is executed. Otherwise, loading `/settings/hosting/:site` on the browser will redirect to `/stats/day` since the plan has not been retrieved (making `isSiteOnAtomicPlan` to return false).